docs: update Jupyter tutorial

alexjuda opened the pull request:
# The problem

1. Our Jupyter tutorial didn't list pre-requirements in the intro.
2. The code snippets weren't backed by tests.
3. The tutorial content was out of date wrt. current capabilities.

# This PR's solution

* Reworks the Jupyter tutorial.
* Moves some testing machinery to under `src` so we can reuse it between standard `/tests` and `/docs/examples/tests`.

# Checklist

_Check that this PR satisfies the following items:_

- [x] Tests have been added for new features/changed behavior (if no new features have been added, check the box).
- [x] The [changelog file](CHANGELOG.md) has been updated with a user-readable description of the changes (if the change isn't visible to the user in any way, check the box).
- [x] The PR's title is prefixed with `<feat/fix/chore/imp[rovement]/int[ernal]/docs>[!]:`
- [x] The PR is linked to a JIRA ticket (if there's no suitable ticket, check the box).

alexjuda committed:
Rework Jupyter tutorial

Add pre-requisite

Denote what file the snippet is supposed to be in

Update ray instructions

Update information about imports for CE

Describe code reuse

Move snippets to test file

Add test for task in another file

Move code reuse paragraph to a note

Add test for wf in another file

Edit changelog

Remove outdated file

Use green-red admonitions for file types

BenjaminMummery commented:
I think this probably addresses https://zapatacomputing.atlassian.net/browse/ORQSDK-898?atlOrigin=eyJpIjoiZDM0OTU2ODUzMjYzNDc3NzgzNTJlY2QxMGU5OTNmZWIiLCJwIjoiaiJ9 as well

BenjaminMummery commented on docs/examples/tests/parsers.py:
```diff
@@ -4,7 +4,7 @@
 
 
-def _get_fn_body_as_str(fn, start_marker: str, end_marker: str) -> str:
+def _get_fn_body_as_str(fn, start_marker: str, end_marker: str, dedent: int) -> str:
```
Praise: this is a really nice addition!

BenjaminMummery commented on docs/tutorials/jupyter-sdk.rst:
```diff
@@ -4,10 +4,16 @@ Using the Workflow SDK with Jupyter
+Prerequisites
+=============
+
+#. You've :doc:`installed Orquestra Workflow SDK<installing-macos-linux>`.
```
Query: do we want to link the windows installation instructions here as well?

alexjuda replied:
So far we only linked to the macOS/Linux installation in our tutorials, because that's the main use case, and the macOS/Linux installation contains a link to the Windows one. I usually prefer conciseness over comprehensiveness.

BenjaminMummery commented on docs/tutorials/jupyter-sdk.rst:
```diff
@@ -19,172 +25,117 @@ Jupyter can then be run using simply
     jupyter notebook
 
-Jupyter Notebooks and the Orquestra Workflow SDK must be installed in the same environment.
+Jupyter Notebooks and the Orquestra Workflow SDK must be installed in the same virtual environment.
```
Suggestion: I'd either use "virtual environment" or "environment" throughout for clarity

alexjuda replied:
Good catch 🤦 a39cdf893d014cf2eb7d8aff349e63b4577cd8ff

BenjaminMummery reviewed (approved):
LGTM, I had a couple of non-blocking suggestions

alexjuda committed:
Unify venv mention

alexjuda committed:
Fix import sorts

alexjuda committed:
Attempt to fix module missing errors in tests

Removes the `PYTHONPATH` override. I don't know if it's required
anymore. Chances are we needed this in the past, but not anymore.

Fixes `ModuleNotFoundError` when running tests with scripts and package
dirs as fixtures. Example:
https://github.com/zapatacomputing/orquestra-workflow-sdk/actions/runs/5877684369/job/15938259250#step:6:2197
